### ISSUE TITLE
Modularize `fedimint-cli` commands

### ISSUE THREAD

**Issue by elsirion** [fedimint/fedimint#3630]
A lot of wallet functionality is still implemented "globally" in `fedimint-cli`, but this gives default modules a weird, special privilege while not allowing working with multiple instances of the same module. Instead all functionality should be made available via the `handle_cli_command` API:

https://github.com/fedimint/fedimint/blob/d9c007d1ef3c9f265ed6a76800081736c7805aa5/fedimint-client/src/module/mod.rs#L250-L254

Before implementing parsing manually we should look into using `clap` in modules. But we need to be cautious not to panic from inside the module since the errors would be hard to interpret, instead passing a clap error up to the client will allow us to add more context.

**Comment by dpc** [comment]
Context: originally there was just `fedimint-cli ...` and each Fedimint was hardcoded to have 3 modules. Fedimint-cli was written around it.

The modularization happened and every Fedimint can set up arbitrary set of modules, even ones Fedimint developers have no idea about. Also Mints can "install" the same module twice, possibly in a different version or with different config, which might come handy sometimes.

The `fedimint-cli`'s top-level commands work as before, by looking up the first (in order of ids) module of a given type and using if (see `get_first_module`). Which 99% of the time is the right thing to do.

Now `fedimint-cli` can't have support for the modules that are not part of Fedimint repo. For these modules there's a `fedimint-cli module <kind/id> ...` which is forwarded to `handle_cli_command` so 3rd party modules can expose and handle arbitrary CLI functionality.

But just like 3rd party modules, the standard modules should be able to handle `handle_cli_command` calls. So if someone having 2 LN modules in their Federation can request a n ln invoice from the second one.

The existing global `fedimint-cli` commands functionality should be kept, but possibly better refactored to become strictly `get_first_module` + method call like, maybe even with some options to change the modules being used.

**Comment by elsirion** [comment]
> The existing global `fedimint-cli` commands functionality should be kept, but possibly better refactored to become strictly `get_first_module` + method call like, maybe even with some options to change the modules being used.

I think `fedimint-cli` would become much more maintainable by only supporting the modularized commands. By putting all client/module commands in one sub-command we can also be more granular about which sub-commands require which arguments (e.g. `--data-dir` for client, but API calls can take either a data dir or a direct URL, …).

How I imagine the command structure:

```
fedimint-cli
├── client
│   ├── join
│   ├── info
│   ├── ...
│   ├── module (to directly address modules by name or id)
│   ├── mint (any unknown command is evaluated as module name)
│   │   ├── spend
│   │   ├── reissue
│   │   └── ...
│   ├── wallet
│   ├── lightning
│   └── ...
├── admin
│   ├── setup (CLI tool to drive initial setup/DKG)
│   ├── status
│   └── audit
└── dev
    ├── api
    │   └── <method> <args...>
    └── ... (maybe even dbtool, recoverytool?)
```

**Comment by mayrf** [comment]
Thanks for the clarifications! I think I sort of have an idea now and would like to try my luck on this issue

**Comment by elsirion** [comment]
This will be very painful if we actually remove the old default module global commands (but I still think we should) because we'd need to rewrite all the devimint tests. So it might be a good idea to first put in the `fedimint-cli client <module_kind> <module_command> <args…>` path and then removing the global commands one by one.

**Comment by dpc** [comment]
> I think `fedimint-cli` would become much more maintainable by only supporting the modularized commands.

I would expect that the global commands (would/already) include at least some cross-module ones, that don't neatly fit into single module api.

Also it seems to me that some most commonly used apis being in a global namespace is just ... convenient. 99% of users will not have multiple instances of these modules. If the whole handler is just `get_first_module<MintClientModule>.do_xyz()` I don't think it's a huge burden to keep such handler.

**Comment by elsirion** [comment]
There are legitimate cross-module ones like balance, join-federation, config, ... These would obviously stay. But e.g. spend and reissue are part of the mint module and when someone creates a mintv2 module for their mint then they couldn't even remove the old one from the client because the client still depends on the old module type.

Is it that bad to call `fedimint-cli client mint spend 42000`? Especially with auto completion that even makes it easier to find stuff if one knows the module but not the exact command name. I don't see many people using the CLI directly anyway and for mprocs we can add aliases if needed.

**Comment by dpc** [comment]
> Is it that bad to call `fedimint-cli client mint spend 42000`

Not too bad, but worse than `fedimint-cli spend 42000`. And we haven't even seen a single Federation with two mint modules yet. :)

> Especially with auto completion that even makes it easier to find stuff if one knows the module but not the exact command name

The auto-completion for module parts of `fedimint-cli` might be tricky because it crosses the two levels of multiple separate clap parsers. I'm not even sure if it's possible ATM.

**Comment by elsirion** [comment]
> And we haven't even seen a single Federation with two mint modules yet. :)

It's less about having two modules of the same kind and more about the inability to fully replace default modules and having to compile in modules that one doesn't want to use anymore.

One thing I'd want to do anyway though is putting all the client commands in their own `client` namespace, i.e. `fedimint-cli client --data-dir=/tmp/foo spend 42000`. That way it will be much easier to tell users which arguments are required where. E.g. the `dev api` call shouldn't require a config but should be able to accept raw websocket URLs imo.

> The auto-completion for module parts of fedimint-cli might be tricky because it crosses the two levels of multiple separate clap parsers. I'm not even sure if it's possible ATM.

Good point :/

**Comment by dpc** [comment]
> It's less about having two modules of the same kind and more about the inability to fully replace default modules and having to compile in modules that one doesn't want to use anymore.

The current "get first module of a kind by instance id" is simplistic and provisional. I think eventually long living Federations will have consensus settings to let clients know which module instances are preferable/obsolete etc, and we'll use `get_preferred_module::<MintClientModule>()` or something smarter. `fedimint-cli send` might be smart enough to prioritize draining obsolete "liquid/cash" (aka primary atm) modules and `fedimint-cli` in general smart enough to avoiding receiving funds to obsolete modules.

So global level commands should always be able to do more or less the right thing.

>  and having to compile in modules that one doesn't want to use anymore

I don't get this part, but hopefully smarter module selecting addresses that as well.

> One thing I'd want to do anyway though is putting all the client commands in their own client namespace,

Usability wise seems weird because `fedimint-cli` is kind of all about client ...

The data-dir being sometimes needed and sometimes not is a bit of a thorn, but I don't think it's a big deal worth making everything slower to type. The current way of making it globally optional seems good enough to me.


In general I would like for `fedimint-cli` in addition to be useful for scripting and automation, also be a decent human-focused CLI client with convenient UX, because .. why not? If it's not someone will have to write one anyway, because people like me will want it.

**Comment by elsirion** [comment]
Yeah, I think our differences here come from what we think `fedimint-cli` is/should be. I see it as a swiss army knife for all things fedimint that could potentially be useful for debugging and testing while you are more focused on the end-user client functionality.

Maybe we should actually have both? We could even do the old `coreutils` trick of having one binary and a bunch of symlinks to it: when it detects being called with a specific name it changes its behavior.

We could have a `fmcli` (no hyphen, shorter) for end users that acts like the current one without the admin commands, maybe even with a default location for wallets in some XDG directory and other amenities, while `fedimint-cli` is more explicit and a debug/testing tool.

**Comment by dpc** [comment]
We can do that, but then why would you use the "bare" one if the "rich" one can do anything that the "bare" one plus more.

> We could have a fmcli (no hyphen, shorter) for end users that acts like the current one without the admin commands

I'd like to have admin commands in my end user wallet too! Admins are users too. :)



> In this case `fedimint-cli` is what one can easily port to custom versions of FM, while in `fmcli` one either needs to fork it or modules are a bit of second class citizens.

I don't understand this one. In the end-user-one there's no problem exposing exactly same per-module cli handling for users that need it, and same 3rd party module injection.

Even for 3rd party modules, thanks to useful abstractions (like an existing "primary module"), even the smart commands can utilize set of common APIs for modules they don't know anything else about.

E.g. `balance` can calculate and add up balances reported by every module. `ln-pay` can use any module that can produce liquid hash. `backup` can backup any module.

In fact that's probably that's a good argument to maintain a "smart end-user CLI command" internally in-tree in the first place - as it forces our design to consider good generalizations that allow generalized operations.

> maybe even with a default location for wallets in some XDG directory and other amenities,

If they were two separate command than I guess I'd pursue it, yes. If they're together I think the env var for `--data-dir` is good enough.

Ideally I'd want by default everything go to `~/.share/fedimint/` and user be able to join multiple federations, give them pet names, set a default one or use a custom via cli argument, and even have some commands that aggregate over them (e.g. show my balance in all federations).

Still, even in that most elaborated visions, I'd like this cli to support everything that the 'bare' client can do, and be usable for scripting (by e.g. just overriding data dir env variable).

**Comment by elsirion** [comment]
TODO: check which functions don't have a module version yet. I think we are nearly there.

**Comment by elsiribot** [comment]
The core modularization work is done:
- All 12 module-specific commands have `module <kind> <command>` equivalents (#8340)
- `config` and `session-count` moved to `dev` subcommand (#8340)
- All hidden global wrappers emit deprecation warnings (#8416)
- Remaining non-hidden globals (`Info`, `Backup`, `Restore`, `PrintSecret`, `ListOperations`, `DiscoverVersion`) are legitimately cross-module / client-level, not module-specific

Remaining follow-up split into:
- #8644 — migrate devimint tests off the deprecated commands (gated on backward-compat window)
- #8645 — remove the deprecated wrappers (blocked on #8644 + deprecation window)

Closing this in favor of the two follow-ups.
